examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,360 @@
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
+import io.circe.{Decoder, Encoder, parser}
+import io.circe.syntax.EncoderOps
+import org.tessellation.currency.dataApplication.dataApplication.DataApplicationValidationErrorOr
+import org.tessellation.currency.dataApplication.{DataApplicationValidationError, DataState, DataUpdate, L1NodeContext}
+import org.tessellation.security.signature.Signed
+import cats.syntax.all._
+import com.my.currency.shared_data.Combiners.{combineCreatePoll, combineVoteInPoll}
+import com.my.currency.shared_data.Data.{CreatePollUpdate, Poll, PollState, PollUpdate, VoteInPollUpdate, State}
+import com.my.currency.shared_data.Validations.{createPollValidationsWithSignature, createPollValidations, voteInPollValidationsWithSignature, voteInPollValidations}
+import com.my.currency.shared_data.TypeValidators.{validateIfOptionExists, validateIfPollAlreadyExists, validateIfUserAlreadyVoted, validateIfVotePollExists, validatePollSnapshotInterval, validateProvidedAddress, validateSnapshotCreatePoll}
+import com.my.currency.shared_data.Utils.{customStateDeserialization, customStateSerialization, customUpdateDeserialization, customUpdateSerialization}
+import monocle.syntax.all._
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.schema.address.Address
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.signature.SignatureProof
+
+import java.nio.charset.StandardCharsets
+
+object Errors {
+  case object PollAlreadyExists extends DataApplicationValidationError {
+    val message = "Poll already exists"
+  }
+
+  case object PollDoesNotExists extends DataApplicationValidationError {
+    val message = "Poll does not exists"
+  }
+
+  case object InvalidOption extends DataApplicationValidationError {
+    val message = "Invalid option"
+  }
+
+  case object InvalidAddress extends DataApplicationValidationError {
+    val message = "Provided address different than proof"
+  }
+
+  case object RepeatedVote extends DataApplicationValidationError {
+    val message = "This user already voted!"
+  }
+
+  case object InvalidEndSnapshot extends DataApplicationValidationError {
+    val message = "Provided end snapshot ordinal lower than current snapshot!"
+  }
+
+  case object CouldNotGetLatestCurrencySnapshot extends DataApplicationValidationError {
+    val message = "Could not get latest currency snapshot!"
+  }
+
+  case object CouldNotGetLatestState extends DataApplicationValidationError {
+    val message = "Could not get latest state!"
+  }
+
+  case object ClosedPool extends DataApplicationValidationError {
+    val message = "Pool is closed"
+  }
+}
+
+object TypeValidators {
+  def validateIfPollAlreadyExists(maybeState: Option[PollState]): DataApplicationValidationErrorOr[Unit] = {
```

Comment:
This validation function and the one below it are a little confusing in that they're only ever dealing with the same option type. It might make more sense to implement this logic:

`validateIfPollAlreadyExists(state.polls.get(voteId.toString))`

from the invocation call of this, directly in the validate function by passing State and voteId as parameters. Since this isn't used anywhere else (i.e. the result of the get is not used anywhere else,) it makes sense to implement it only in the validator.

Reply:
Done